go/libraries/doltcore/doltdb/doltdb.go

```diff
@@ -1457,6 +1508,57 @@ func (ddb *DoltDB) CommitWithWorkingSet(
 	return NewCommit(ctx, ddb.vrw, ddb.ns, dc)
 }
 
+// writeWorkingSetAndNotifyListeners writes the specified |workingSet| at the specified |workingSetRef| with the
+// specified ws metadata, |meta|, in the dataset |wsDs| and returns the created WorkingSetSpec along with any error
+// encountered. If any listeners are registered for working root updates, then they will be notified as well.
+func (ddb *DoltDB) writeWorkingSetAndNotifyListeners(ctx context.Context, workingSetRef ref.WorkingSetRef, workingSet *WorkingSet, meta *datas.WorkingSetMeta, wsDs datas.Dataset) (wsSpec *datas.WorkingSetSpec, err error) {
+	var prevRoot RootValue
+	if wsDs.HasHead() {
+		prevWorkingSet, err := newWorkingSet(ctx, workingSetRef.String(), ddb.vrw, ddb.ns, wsDs)
+		if err != nil {
+			return nil, err
+		}
+		prevRoot = prevWorkingSet.workingRoot
+	} else {
+		// If the working set dataset doesn't have a head, then there isn't a previous root value for us to use
+		// for the database update, so instead we pass in an EmptyRootValue so that implementations of
+		// DatabaseUpdateListener don't have to do nil checking. This can happen when a new database is created
+		// or when a new branch is created.
+		prevRoot, err = EmptyRootValue(ctx, ddb.vrw, ddb.ns)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var branchName string
+	if strings.HasPrefix(workingSet.Name, "heads/") {
+		branchName = workingSet.Name[len("heads/"):]
+	}
+
+	wsSpec, err = workingSet.writeValues(ctx, ddb, meta)
+	if err != nil {
+		return nil, err
+	}
+
+	if branchName != "" {
+		for _, listener := range DatabaseUpdateListeners {
+			sqlCtx, ok := ctx.(*sql.Context)
+			if ok {
+				err := listener.WorkingRootUpdated(sqlCtx,
```

Comment:
I initially had it in the `doltdb` package because of the closeness to types like `RootValue` that it's using and because functions in `doltdb` like `writeSession` need to reference the UpdateListener type. 

I just tried moving this type to the `sqle` namespace, but it results in import cycles since the `doltdb` package needs to access the list of registered update listeners as part of writing the session.